EIPS/eip-3508.md

```diff
@@ -0,0 +1,100 @@
+
+The purpose of this EIP is to render these features viable by introducing a way via which multi-contract systems are able to access the same in-memory data source without necessarily transmitting the full payload between them.
+
+This EIP enables elaborate smart contract features to become part of a larger call-chain by efficiently reading data from the original transaction payload rather than requiring the data to be passed in as call-level data. Its inclusion will mainly benefit advanced trustless schemes to manifest, such as efficient verification of Merkle Patricia trees validating the storage value of a particular Ethereum block or EVM-based layer 2 solutions.
+
+A side-effect of this change is that smart contract systems relying entirely on origin data inherently guarantee that the data they receive has not been malformed by an intermediate smart contract call.
+
+## Specification
+
+### ORIGINDATALOAD (`0x47`), ORIGINDATASIZE (`0x48`) and ORIGINDATACOPY (`0x49`)
+
+These instructions are meant to operate identically to their call-prefixed counterparts with the exception that they instead operate on the original `data` of a transaction instead of the current call's data. In detail:
+
+- ORIGINDATALOAD (`0x47`) performs identically to CALLDATALOAD (`0x35`)
+- ORIGINDATASIZE (`0x48`) performs identically to CALLDATASIZE (`0x36`)
+- ORIGINDATACOPY (`0x49`) performs identically to CALLDATACOPY (`0x37`)
+
+As the data is retrieved once again from the execution environment, the costs for the three instructions will be `G_verylow`, `G_base` and `G_base + G_verylow * (number of words copied, rounded up)` respectively.
+
+### `AUTHCALL` (`0xf7`) Interaction
+
+The [EIP-3074](./eip-3074) that will be part of the London fork has introduced a new call instruction called `AUTHCALL` (`0xf7`) that will replace a transaction's `ORIGIN` (`0x32`) with the context variable `authorized`. The intention of `AUTHCALL` is to prevent discrimination between smart contracts and EOAs which `ORIGIN` initially facilitated and as a result, it is sensible also replace the values retrieved by the `ORIGINDATA*` opcodes to the ones used in the `AUTHCALL`.
+
+In detail, the transaction data the `ORIGINDATA*` opcodes operate on will be equivalent to the `calldata` specified in the `commit` of the `AUTH` message format. If a single-call `AUTH` message is defined, the opcodes will retrieve the data of `calldata`. Otherwise, the opcodes will retrieve the `calldata` of each sub-call (i.e. `call 1`, `call 2` ... `call X`) depending on the call execution context.
```

Comment:
```suggestion
The transaction data the `ORIGINDATA*` opcodes operate on will be equivalent to the `args*` parameter to the nearest `AUTHCALL` up the stack.  If there is no `AUTHCALL` in the stack then `ORIGINDATA*` will operate on the transaction's data field.
```
I'm pretty sure this is what you mean here.  Using the `AUTH` commit would be very bizarre.